homeassistant/components/prowl/notify.py

```diff
@@ -33,46 +34,49 @@ async def async_get_service(
+    prowl = await hass.async_add_executor_job(
+        partial(prowlpy.Prowl, apikey=config[CONF_API_KEY])
+    )
+    return ProwlNotificationService(hass, prowl)
 
 
 class ProwlNotificationService(BaseNotificationService):
     """Implement the notification service for Prowl."""
 
-    def __init__(self, hass, api_key):
+    def __init__(self, hass: HomeAssistant, prowl: prowlpy.Prowl) -> None:
         """Initialize the service."""
         self._hass = hass
-        self._api_key = api_key
+        self._prowl = prowl
 
-    async def async_send_message(self, message, **kwargs):
+    async def async_send_message(self, message: str, **kwargs: Any) -> None:
         """Send the message to the user."""
-        response = None
-        session = None
-        url = f"{_RESOURCE}add"
-        data = kwargs.get(ATTR_DATA)
-        payload = {
-            "apikey": self._api_key,
-            "application": "Home-Assistant",
-            "event": kwargs.get(ATTR_TITLE, ATTR_TITLE_DEFAULT),
-            "description": message,
-            "priority": data["priority"] if data and "priority" in data else 0,
-        }
-        if data and data.get("url"):
-            payload["url"] = data["url"]
-
-        _LOGGER.debug("Attempting call Prowl service at %s", url)
-        session = async_get_clientsession(self._hass)
+        data = kwargs.get(ATTR_DATA, {})
+        if data is None:
+            data = {}
 
         try:
             async with asyncio.timeout(10):
-                response = await session.post(url, data=payload)
-                result = await response.text()
-
-            if response.status != HTTPStatus.OK or "error" in result:
-                _LOGGER.error(
-                    "Prowl service returned http status %d, response %s",
-                    response.status,
-                    result,
+                await self._hass.async_add_executor_job(
+                    partial(
+                        self._prowl.send,
+                        application="Home-Assistant",
+                        event=kwargs.get(ATTR_TITLE, ATTR_TITLE_DEFAULT),
+                        description=message,
+                        priority=data.get("priority", 0),
+                        url=data.get("url"),
+                    )
                 )
-        except TimeoutError:
-            _LOGGER.error("Timeout accessing Prowl at %s", url)
+        except TimeoutError as ex:
+            _LOGGER.error("Timeout accessing Prowl API")
+            raise HomeAssistantError("Timeout accessing Prowl API") from ex
+        except prowlpy.APIError as ex:
+            if str(ex).startswith("Invalid API key"):
+                _LOGGER.error("Invalid API key for Prowl service")
+                raise HomeAssistantError("Invalid API key for Prowl service") from ex
+            if str(ex).startswith("Not accepted"):
+                _LOGGER.error("Prowl returned: exceeded rate limit")
```

Comment:
Please don't log an error if also raising an exception. Let the handler of the exception decide if it should log the message.